File: lib/features/assistant/domain/local_intent_parser.dart
import '../../cycle/domain/cycle_models.dart';
import '../../reminders/domain/reminder_models.dart';

enum LocalIntent {
  periodStarted,
  periodEnded,
  logFlow,
  logSymptom,
  addReminder,
  nextPeriodQuery,
  historyQuery,
  unknown,
}

class ParsedCommand {
  const ParsedCommand({
    required this.intent,
    required this.requiresConfirmation,
    this.date,
    this.value,
    this.rawText,
    this.reminderHour,
    this.reminderMinute,
    this.reminderKind,
  });

  final LocalIntent intent;
  final bool requiresConfirmation;
  final DateTime? date;
  final String? value;
  final String? rawText;
  final int? reminderHour;
  final int? reminderMinute;
  final ReminderKind? reminderKind;

  ObservationKind? get observationKind {
    if (intent == LocalIntent.logFlow) return ObservationKind.menstrualFlow;
    if (intent != LocalIntent.logSymptom) return null;
    final text = (value ?? '').toLowerCase();
    const mapping = <String, ObservationKind>{
      'cramp': ObservationKind.cramps,
      'migraine': ObservationKind.migraine,
      'headache': ObservationKind.headache,
      'back pain': ObservationKind.backPain,
      'breast': ObservationKind.breastTenderness,
      'bloat': ObservationKind.bloating,
      'acne': ObservationKind.acne,
      'nausea': ObservationKind.nausea,
      'digest': ObservationKind.digestion,
      'fatigue': ObservationKind.fatigue,
      'dizz': ObservationKind.dizziness,
      'appetite': ObservationKind.appetite,
      'craving': ObservationKind.cravings,
      'sleep': ObservationKind.sleep,
      'energy': ObservationKind.energy,
      'stress': ObservationKind.stress,
      'mood': ObservationKind.mood,
      'anxiety': ObservationKind.anxiety,
      'irrit': ObservationKind.irritability,
      'libido': ObservationKind.libido,
    };
    for (final entry in mapping.entries) {
      if (text.contains(entry.key)) return entry.value;
    }
    return ObservationKind.custom;
  }

  FlowLevel? get flowLevel => switch ((value ?? '').toLowerCase()) {
    'spotting' => FlowLevel.spotting,
    'light' => FlowLevel.light,
    'medium' => FlowLevel.medium,
    'heavy' => FlowLevel.heavy,
    _ => null,
  };

  ObservationSeverity? get severity {
    final text = (value ?? rawText ?? '').toLowerCase();
    if (text.contains('severe') ||
        text.contains('very bad') ||
        text.contains('bad cramps')) {
      return ObservationSeverity.severe;
    }
    if (text.contains('moderate')) return ObservationSeverity.moderate;
    if (intent == LocalIntent.logSymptom) return ObservationSeverity.mild;
    return null;
  }

  String? get label {
    if (intent == LocalIntent.logFlow) return '${value ?? 'Flow'} flow';
    if (intent == LocalIntent.addReminder)
      return _reminderLabel(value ?? rawText ?? 'Health reminder');
    final kind = observationKind;
    return kind?.name;
  }

  String? get note => rawText;

  String _reminderLabel(String input) {
    var text = input.trim();
    text = text.replaceFirst(
      RegExp(r'^remind me(?: to| about)?\s*', caseSensitive: false),
      '',
    );
    text = text.replaceAll(
      RegExp(
        r'\s+at\s+\d{1,2}(?::\d{2})?\s*(?:am|pm)?\s*$',
        caseSensitive: false,
      ),
      '',
    );
    return text.trim().isEmpty ? 'Health reminder' : text.trim();
  }
}

class LocalIntentParser {
  ParsedCommand parse(String input, {required DateTime now}) {
    final text = input.trim().toLowerCase().replaceAll(RegExp(r'\s+'), ' ');
    final today = DateTime(now.year, now.month, now.day);
    final date = text.contains('yesterday')
        ? today.subtract(const Duration(days: 1))
        : today;

    if (text.contains('period') &&
        _containsAny(text, ['started', 'start', 'began', 'begin'])) {
      return ParsedCommand(
        intent: LocalIntent.periodStarted,
        date: date,
        requiresConfirmation: true,
        rawText: input,
      );
    }
    if (text.contains('period') &&
        _containsAny(text, ['ended', 'end', 'stopped', 'stop'])) {
      return ParsedCommand(
        intent: LocalIntent.periodEnded,
        date: date,
        requiresConfirmation: true,
        rawText: input,
      );
    }
    for (final flow in ['spotting', 'light', 'medium', 'heavy']) {
      if (text.contains(flow) &&
          _containsAny(text, ['flow', 'bleeding', 'yesterday', 'today'])) {
        return ParsedCommand(
          intent: LocalIntent.logFlow,
          date: date,
          value: flow,
          requiresConfirmation: true,
          rawText: input,
        );
      }
    }
    if (_containsAny(text, [
      'cramp',
      'headache',
      'migraine',
      'back pain',
      'breast',
      'bloating',
      'acne',
      'nausea',
      'fatigue',
      'dizzy',
      'dizziness',
      'appetite',
      'craving',
      'sleep',
      'energy',
      'stress',
      'mood',
      'anxiety',
      'irritable',
      'irritability',
      'libido',
    ])) {
      return ParsedCommand(
        intent: LocalIntent.logSymptom,
        date: date,
        value: text,
        requiresConfirmation: true,
        rawText: input,
      );
    }
    if (text.contains('remind me')) {
      final time = _parseTime(text);
      return ParsedCommand(
        intent: LocalIntent.addReminder,
        date: date,
        value: input,
        reminderHour: time?.$1,
        reminderMinute: time?.$2,
        reminderKind: _reminderKind(text),
        requiresConfirmation: true,
        rawText: input,
      );
    }
    if (_containsAny(text, [
      'next period',
      'when is my period',
      'when should my period',
    ])) {
      return ParsedCommand(
        intent: LocalIntent.nextPeriodQuery,
        requiresConfirmation: false,
        rawText: input,
      );
    }
    if (_containsAny(text, [
      'last six periods',
      'period history',
      'show my periods',
    ])) {
      return ParsedCommand(
        intent: LocalIntent.historyQuery,
        requiresConfirmation: false,
        rawText: input,
      );
    }
    return ParsedCommand(
      intent: LocalIntent.unknown,
      date: date,
      requiresConfirmation: true,
      rawText: input,
    );
  }

  (int, int)? _parseTime(String text) {
    final match = RegExp(
      r'\bat\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?\b',
      caseSensitive: false,
    ).firstMatch(text);
    if (match == null) return null;
    var hour = int.tryParse(match.group(1) ?? '');
    final minute = int.tryParse(match.group(2) ?? '0') ?? 0;
    if (hour == null || hour > 23 || minute > 59) return null;
    final meridiem = match.group(3)?.toLowerCase();
    if (meridiem != null) {
      if (hour < 1 || hour > 12) return null;
      if (meridiem == 'pm' && hour != 12) hour += 12;
      if (meridiem == 'am' && hour == 12) hour = 0;
    }
    return (hour, minute);
  }

  ReminderKind _reminderKind(String text) {
    if (_containsAny(text, ['contraception', 'birth control', 'pill']))
      return ReminderKind.contraception;
    if (text.contains('supplement')) return ReminderKind.supplement;
    if (text.contains('ovulation')) return ReminderKind.ovulationTest;
    if (text.contains('pregnancy test')) return ReminderKind.pregnancyTest;
    return ReminderKind.medication;
  }

  bool _containsAny(String text, List<String> values) =>
      values.any(text.contains);
}

File: web/src/features/assistant/local-intent-parser.ts
import type { FlowLevel, ObservationKind, Severity } from '../../domain/cycle/types';

export const ASSISTANT_CAPABILITY_IDS = [
  'INT-001', 'INT-002', 'INT-003', 'INT-004', 'INT-005',
  'INT-006', 'INT-007', 'INT-008', 'INT-009',
] as const;

export type LocalIntent =
  | 'periodStarted'
  | 'periodEnded'
  | 'logFlow'
  | 'logSymptom'
  | 'addReminder'
  | 'nextPeriodQuery'
  | 'historyQuery'
  | 'searchHistory'
  | 'unknown';

export type ReminderKind =
  | 'medication'
  | 'contraception'
  | 'supplement'
  | 'ovulationTest'
  | 'pregnancyTest';

export type ParsedCommand = {
  intent: LocalIntent;
  requiresConfirmation: boolean;
  date?: string;
  value?: string;
  rawText: string;
  reminderHour?: number;
  reminderMinute?: number;
  reminderKind?: ReminderKind;
  observationKind?: ObservationKind;
  flowLevel?: FlowLevel;
  severity?: Severity;
  label?: string;
  note?: string;
};

const SYMPTOMS: Array<[string, ObservationKind]> = [
  ['cramp', 'cramps'],
  ['migraine', 'migraine'],
  ['headache', 'headache'],
  ['back pain', 'backPain'],
  ['breast', 'breastTenderness'],
  ['bloat', 'bloating'],
  ['acne', 'acne'],
  ['nausea', 'nausea'],
  ['digest', 'digestion'],
  ['fatigue', 'fatigue'],
  ['dizz', 'dizziness'],
  ['appetite', 'appetite'],
  ['craving', 'cravings'],
  ['sleep', 'sleep'],
  ['energy', 'energy'],
  ['stress', 'stress'],
  ['mood', 'mood'],
  ['anxiety', 'anxiety'],
  ['irrit', 'irritability'],
  ['libido', 'libido'],
];

function dateKey(date: Date): string {
  return [
    date.getUTCFullYear(),
    String(date.getUTCMonth() + 1).padStart(2, '0'),
    String(date.getUTCDate()).padStart(2, '0'),
  ].join('-');
}

function localDateFor(text: string, now: Date): string {
  const date = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate()));
  if (text.includes('yesterday')) date.setUTCDate(date.getUTCDate() - 1);
  return dateKey(date);
}

function containsAny(text: string, values: string[]): boolean {
  return values.some((value) => text.includes(value));
}

function parseTime(text: string): [number, number] | null {
  const match = /\bat\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?\b/i.exec(text);
  if (!match) return null;
  let hour = Number.parseInt(match[1], 10);
  const minute = Number.parseInt(match[2] ?? '0', 10);
  if (!Number.isFinite(hour) || hour > 23 || minute > 59) return null;
  const meridiem = match[3]?.toLowerCase();
  if (meridiem) {
    if (hour < 1 || hour > 12) return null;
    if (meridiem === 'pm' && hour !== 12) hour += 12;
    if (meridiem === 'am' && hour === 12) hour = 0;
  }
  return [hour, minute];
}

function reminderKind(text: string): ReminderKind {
  if (containsAny(text, ['contraception', 'birth control', 'pill'])) return 'contraception';
  if (text.includes('supplement')) return 'supplement';
  if (text.includes('ovulation')) return 'ovulationTest';
  if (text.includes('pregnancy test')) return 'pregnancyTest';
  return 'medication';
}

function reminderLabel(input: string): string {
  let text = input.trim().replace(/^remind me(?: to| about)?\s*/i, '');
  text = text.replace(/\s+at\s+\d{1,2}(?::\d{2})?\s*(?:am|pm)?\s*$/i, '');
  return text.trim() || 'Health reminder';
}

function symptomKind(text: string): ObservationKind {
  for (const [needle, kind] of SYMPTOMS) {
    if (text.includes(needle)) return kind;
  }
  return 'custom';
}

function symptomSeverity(text: string): Severity {
  if (containsAny(text, ['severe', 'very bad', 'bad cramps'])) return 'severe';
  if (text.includes('moderate')) return 'moderate';
  return 'mild';
}

export function parseLocalIntent(input: string, now: Date): ParsedCommand {
  const text = input.trim().toLowerCase().replace(/\s+/g, ' ');
  const date = localDateFor(text, now);

  if (text.includes('period') && containsAny(text, ['started', 'start', 'began', 'begin'])) {
    return { intent: 'periodStarted', date, requiresConfirmation: true, rawText: input, note: input };
  }

  if (text.includes('period') && containsAny(text, ['ended', 'end', 'stopped', 'stop'])) {
    return { intent: 'periodEnded', date, requiresConfirmation: true, rawText: input, note: input };
  }

  for (const flow of ['spotting', 'light', 'medium', 'heavy'] as const) {
    if (text.includes(flow) && containsAny(text, ['flow', 'bleeding', 'yesterday', 'today'])) {
      return {
        intent: 'logFlow',
        date,
        value: flow,
        flowLevel: flow,
        observationKind: 'menstrualFlow',
        label: `${flow} flow`,
        requiresConfirmation: true,
        rawText: input,
        note: input,
      };
    }
  }

  const searchMatch = /^(?:search|find|show)\s+(?:my\s+)?(?:history\s+for\s+)?(.+)$/.exec(text);
  if (searchMatch && searchMatch[1]) {
    return { intent: 'searchHistory', value: searchMatch[1].trim(), requiresConfirmation: false, rawText: input };
  }

  if (SYMPTOMS.some(([needle]) => text.includes(needle))) {
    const observationKind = symptomKind(text);
    return {
      intent: 'logSymptom',
      date,
      value: text,
      observationKind,
      severity: symptomSeverity(text),
      label: observationKind,
      requiresConfirmation: true,
      rawText: input,
      note: input,
    };
  }

  if (text.includes('remind me')) {
    const time = parseTime(text);
    return {
      intent: 'addReminder',
      date,
      value: input,
      reminderKind: reminderKind(text),
      ...(time ? { reminderHour: time[0], reminderMinute: time[1] } : {}),
      label: reminderLabel(input),
      requiresConfirmation: true,
      rawText: input,
      note: input,
    };
  }

  if (containsAny(text, ['next period', 'when is my period', 'when should my period'])) {
    return { intent: 'nextPeriodQuery', requiresConfirmation: false, rawText: input };
  }

  if (containsAny(text, ['last six periods', 'period history', 'show my periods'])) {
    return { intent: 'historyQuery', requiresConfirmation: false, rawText: input };
  }

  return { intent: 'unknown', date, requiresConfirmation: true, rawText: input, note: input };
}
